Feat/Add documentation about active/active not supported and possible workarounds (#17)
3scale/3scale-Operations: docs/ha_3scale.adoc

.. AWS RDS: *System* and *Zync*
.. AWS ElastiCaches: *Backend* and *System*
. Execute step 4 and 5 from <<manual-failover-shared-databases>>

=== Active-Active clusters (not supported)

Having an application working on *active/active* configuration is always difficult for any application:

- *Soft Databases limitation*: it is complex but can be achieved with some trade offs, but it is mainly responsibility from the administrator (not 3scale responsibility)
- *Hard Application limitation*: this one is a hard limitation imposed by how 3scale manages OpenShift Routes and cannot be achieved

==== Soft Databases limitation (out of 3scale scope)

One of the most difficult parts for any application deployed on *active*/*active* (not only 3scale), is having databases on *active*/*active* mode, and in case of 3scale there are many different databases involved which would require different *active*/*active* implementations for each of them:

- System-mysql (or system-postgres or system-oracle, you can choose)
- System-redis
- Backend-redis
- Zync-database (postgresql)
- system-sphinx (doesn't need HA)
- system-memcached (doesn't need HA)

> So, from 3scale point of view, an administrator need to ensure that database connection strings configured in 3scale APIManager custom resource on any OCP cluster will be always a master instance (allowing write operations).

Reviewer: Do we know that the db clients we use in 3scale are compatible with the active/active soluitons these DB offerings provide? For example MySQL's Galera cluster or MariaDB and postgres crucnhy DB.

Obviously for Redis we had to upgrade our redis clients to support redis sentinels and redis enterprise but maybe the same is not true for the other DBs we support and their respective offerings that support Active/Active.

Author: I have checked that most of this A/A DB solutions like mysql galera o postgres cruncyDB:
- Can be accessed with smart db clients by adding your database instances on the connection string (possibly not supported on db clients used by 3scale)
- Can deploy official TCP smart load balancers (galera LB, haproxy, nginx...) in front of DB nodes, so there is a single shared LB across your DB nodes that will choose the less loaded backend DB node (less connections...)... So from 3scale point of view any standard db client should work by providing a single DB hostname (the Load Balancer)

> So, from 3scale point of view, an administrator need to ensure that database connection strings configured in 3scale APIManager custom resource on any OCP cluster will be always a master instance (permitting write operations).

Reviewer 2: ```suggestion
So, from 3scale point of view, an administrator need to ensure that database connection strings configured in 3scale APIManager custom resource on any OCP cluster will be always a master instance (allowing write operations).
```

____
*NOTE*

> *This database limitation is a soft limitation because although being very complex (implementing active/active minimum for mysql, postgresql, redis is not easy), can be achieved.*

Reviewer 2: ```suggestion
*This database limitation is a soft limitation because although being very complex (implementing active/active for mysql, postgresql or redis is not trivial), it can be achieved.*
```

____

> ==== Hard Application limitation: OpenShift Routes

Reviewer: is this the only hard limitation we have in the whole 3scale logic that makes supporting Active/Active difficult (ignoring the DB active/active complexities for now)?

Author: This is the only one I thought, but everything  written here is theoretical, that's why I preferred to ask you in case you might think about other possible limitations knowing the things you have seen in support team done by on-prem customers

> However, the main reason why *active*/*active* can not be done although you previously deployed databases on *active*/*active*, is *System* dev/admin portals and *Apicasts* OpenShift Routes (that need to be the same on all OCP clusters) being managed by *Zync* component.

Reviewer 2: ```suggestion
The main reason why *active*/*active* can not be achieved, even with *active*/*active* databases, is *System* dev/admin portals and *Apicasts* OpenShift Routes. They are  managed by the *Zync* component and they need to be the exactly the same on all OCP clusters.
```

. System-app enqueues the background job to create those OpenShift Routes into system-redis
. System-sidekiq takes the job from system-redis, process it and contacts with zync API (zync deployment)
. Zync API (zync deployment) creates a background job into zync-database (postgres)
. Zync-que takes the route-creation-job from zync-database (postgres) and create the final OpenShift  Routes into the cluster that proccessed that job

When having N clusters, imagine Cluster-A and Cluster-B, any of them can receive traffic managed by either Cluster-A-system-app or Cluster-B-system-app, so they will enqueue the job into its own database, and would follow the whole procedure:

```
System-app -> system-redis -> system-sidekiq -> zync (api) -> zync-database -> zync-que -> OpenShift Routes creation
```

And finally, the OpenShift Routes will be created *only* into the cluster that processed the route-creation-job (not on both OCP clusters), so incoming traffic managed by OpenShift Router will work only on one cluster (the one with the needed OpenShift Routes), so not having *active*/*active*.

____
*NOTE*

> *This application limitation is a hard limitation because there is no way to solve it, since the OpenShift Routes creation need to be done on any OCP cluster working on active/active, but this OpenShift Routes creation will be done just on a single cluster.*

Reviewer 2: ```suggestion
*This application limitation is a hard limitation because the first worker that fetches the job, will create the router on the cluster it's running. So, each cluster will have a subset of routes. *
```

____

=== Active independent clusters

A few customers wanting *active*/*active*, as it can not be achieved within the same 3scale instance due to the hard limitation imposed by the OpenShift Routes created by *Zync*, what they have done is to deploy 3scale independent instances on different OCP clusters, where every 3scale instance has its own databases, which are not shared and not synced between them.

Here you can find some considerations:

- Each 3scale independent instance will need its own High Availability configuration
- You need to manage the 3scale configuration with a shared CI/CD, to ensure that any independent 3scale instance will have exactly the same configuration
- You need a kind of Global Load Balancer (maybe done at DNS level above all OCP clusters), to ensure that the same % of traffic goes to every independent 3scale instance on its OCP cluster (maybe using round-robin policy)

> - The rate limit will be independent for every 3scale instance, so if for example you have 2 3scale independent instances, and want a global rate limit of 100 requests/minute, you will need to configure a rate limit of 50 requests/minute on each 3scale instance, and in case of having round-robin DNS policy on the Global Load Balancer, *theoretically* you will achieve the supposed global 100 requests/minute rate limit

Reviewer: Wouldn't there also be the consideration if RH SSO is integrated for OIDC on the API traffic then duplication of clients across the multiple 3scale instances (so the same token can be authorised in any 3scale) would then cause synchronisation issues and ultimately data inconsistencies? Seems like another zync hard limitation maybe?

Author: Mmmm interesting, I have no experience about using 3scale with RH SSO integration (on 3cale SaaS we dont provide our own RH SSO, any customer can use their own OIDC solution that will be contacted by Zync-SaaS).

I would like to know more details about this specific possible problem you have seen with 3scale customers and weird HA solutions

Can you please share a bit more detail about it?

Author: Hey @kevprice83 , did you have time to think about the possibility of shared RH SSO between independent 3scale instances (using same or different wildcard domain)??

Reviewer: @slopezz so thinking about the proposal here: independent active 3scale instances not sharing databases, the issue to be careful with when integrating with RH SSO is if the RH SSO is shared between 3scale instances then the shared CI/CD should probably solve that problem. The shared CI/CD would ensure that the state of all objects synchronised between 3scale and RH SSO, via zync, is consistent. 

Regarding the `WILDCARD_DOMAIN` this would not impact the RH SSO integration to my knowledge. Only if RH SSO was not shared (1 per 3scale instance) would they need to take care of the DNS on the RH SSO side so that it was consistent which would then ensure the `iss` claim is always the same and therefore it won't matter which 3scale instance the API request + JWT reaches as it would be valid in both.

I would also expect that in this scenario you are describing the `WILDCARD_DOMAIN` is always the same and the DNS + loadbalancing is handled on the OpenShift networking layer.

Does that sound like it fits with the architecture you had in mind here?

Author: Thanks @kevprice83 for checking it, that's what I had in mind

I have updated a bit the case of possible active independent clusters on commit https://github.com/3scale/3scale-Operations/pull/17/commits/09e5b022594118c011993e2731f76d8aa7ce184c, please check it, if you think it is good enough, please approve the PR and I will merge it
